### PR TITLE
Add Hausdorff loss

### LINKED ISSUE #6993: Hausdorff distance loss support
Hausdorff distance is widely used in evaluating medical image segmentation methods. Adding an objective/loss function directly to optimize this distance can be instrumental in optimizing this score [1].

<center>Hausdorff Distance</center>

An existing implementation is [publicly available on GitHub](https://github.com/PatRyg99/HausdorffLoss) by [Patryk Rygiel](https://github.com/PatRyg99), although it might take some engineering work to integrate within the MONAI framework.

I was able to train a 3D segmentation model successfully using MONAI's SwinUNETR with this implementation and I would like to contribute the code for this loss so that it's more widely available and easier to integrate with MONAI-based repositories.

References:
[1] Karimi, D., & Salcudean, S. E. (2019). [Reducing the Hausdorff distance in medical image segmentation with convolutional neural networks](https://ieeexplore.ieee.org/abstract/document/8767031?casa_token=OGugo3jEuXQAAAAA:y5fp0CuZiEWlbx8FS4XKHu43Wi_W7albLeIAjsnZ1c-IlTpJmEFJUgoMlKl_taSYuBpcPyCvtTMe). IEEE Transactions on medical imaging, 39(2), 499-513.

### PR DESCRIPTION
**Pull request by imadtoubal**
Integrating an existing implementation publicly available on GitHub by Patryk Rygiel into the MONAI framework.

Fixes #6993.
Fixes https://github.com/Project-MONAI/MONAI/issues/3481
Fixes https://github.com/Project-MONAI/MONAI/issues/1039

### Description

Hausdorff distance is widely used in evaluating medical image segmentation methods. Adding an objective/loss function directly to optimize this distance can be instrumental in optimizing this score [1].

<center>Hausdorff Distance</center>

An existing implementation is [publicly available on GitHub](https://github.com/PatRyg99/HausdorffLoss) by [Patryk Rygiel](https://github.com/PatRyg99), although it might take some engineering work to integrate within the MONAI framework.

I was able to train a 3D segmentation model successfully using MONAI's SwinUNETR with this implementation and I would like to contribute the code for this loss so that it's more widely available and easier to integrate with MONAI-based repositories.

**References**:
[1] Karimi, D., & Salcudean, S. E. (2019). [Reducing the Hausdorff distance in medical image segmentation with convolutional neural networks](https://ieeexplore.ieee.org/abstract/document/8767031?casa_token=OGugo3jEuXQAAAAA:y5fp0CuZiEWlbx8FS4XKHu43Wi_W7albLeIAjsnZ1c-IlTpJmEFJUgoMlKl_taSYuBpcPyCvtTMe). IEEE Transactions on medical imaging, 39(2), 499-513.

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [x] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [x] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [x] In-line docstrings updated.
- [x] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

File: tests/test_hausdorff_loss.py [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+import unittest
+
+import numpy as np
+import torch
+from parameterized import parameterized
+
+from monai.losses import HausdorffDTLoss
+
+TEST_CASES = [
+    [  # shape: (1, 1, 2, 2), (1, 1, 2, 2)
```
Comment (wyli) [inline tests/test_hausdorff_loss.py:23]: thanks for the PR, please help add some 3D and GPU test here, for example https://github.com/Project-MONAI/MONAI/blob/281cb0119c01eaa8e6c841880b91f92f45e8d7f7/tests/test_focal_loss.py#L27

File: tests/test_hausdorff_loss.py [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+        {
+            "input": torch.tensor([[[[1.0, -0.0], [-1.0, 1.0]]], [[[1.0, -1.0], [-1.0, 1.0]]]]),
+            "target": torch.tensor([[[[1.0, 1.0], [1.0, 1.0]]], [[[1.0, 0.0], [1.0, 0.0]]]]),
+        },
+        2.661359,
+    ],
+]
+
+
+class TestHausdorffDTLoss(unittest.TestCase):
```
Comment (wyli) [inline tests/test_hausdorff_loss.py:132]: please add a `skipUnless` decorator to skip the `monai.utils.module.OptionalImportError: from scipy.ndimage.morphology import distance_transform_edt`

see some guide here:
https://github.com/Project-MONAI/MONAI/blob/281cb0119c01eaa8e6c841880b91f92f45e8d7f7/CONTRIBUTING.md?plain=1#L211-L222

**Review by wyli — COMMENTED** [review]
thanks!

File: tests/test_hausdorff_loss.py (hunk 1 above)
Comment (imadtoubal) [reply tests/test_hausdorff_loss.py]: Thanks for the feedback, this has been addressed in commits [3163797](https://github.com/Project-MONAI/MONAI/pull/6994/commits/31637978cc817b8c085c7754fcf28f91ae3a515c) and [8c866de](https://github.com/Project-MONAI/MONAI/pull/6994/commits/8c866de1a15b625c8df444df50a71d9c5f08c3f4)

**Review by wyli — APPROVED** [review]
Thanks, it looks good to me, merging if all tests work fine.

**Pull request merged**